src/shogun/base/SGObject.h

```diff
@@ -39,6 +39,30 @@ class Parameter;
 class CSerializableFile;
 class ParameterObserverInterface;
 
+// all shogun base classes for put/add templates
+class CMachine;
+class CKernel;
+class CDistance;
+class CFeatures;
+class CLabels;
+class CECOCEncoder;
+class CECOCDecoder;
+class CMulticlassStrategy;
+
+template <class T>
+struct is_sg_base
```

Comment:
i wonder whether actually we could generate from a constexpr or something as for a while i imagine that this template will be changed....

Reply:
this will be changed definitely.
I dont know exactly what you have in mind there. Have an example?

Reply:
You might specialize is_sg_base for specific `T`.